File: frontend/website/src/Articles/Article.js
import React from 'react'
import './Articles.css'

export default function Article(props) {
    const biases = {'-1': 'L', '-0.5': 'CL', '0': 'C', '0.5': 'CR', '1': 'R'}
    const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
    const formatDate = date => {
        return months[date.getMonth()] + ' ' + date.getDate() + ', ' + 
            date.getFullYear();
    }
    const capitalize = text => 
        (text[0].toUpperCase() + text.slice(1, text.length))
    const getSentimentClass = () => {
        let sentiment = props.article.textInfo.sentiment;
        let className = 'article-sentiment ';
        if (sentiment > 0) 
            className += 'positive'
        else if (sentiment < 0)
            className += 'negative'
        return className
    }
    const getEmotionClass = () => {
        let sentiment = props.article.textInfo.emotion;
        let className = 'article-sentiment ';
        className += sentiment === 'joy' ? 'positive' : 'negative'
        return className
    }


    let image = <span />;
    if (props.article.imageUrl !== "")
        image = <div className="article-image">
                    <img
                        src={props.article.imageUrl} 
                        alt="news"
                    />
                </div>

    let bias = biases[props.article.company.bias.toString()]
    return (
        <a href={props.article.articleUrl} className="article">
            <div className="left-container">
                <div className="image-title-container">
                    {image}
                    <div className="article-title">
                        {props.article.title}
                    </div>
                </div>
                <div className="article-description">
                    {props.article.description}
                </div>
                <div className="company-date-container">
                    <span className="article-company">
                        {props.article.company.name + ' '}
                    </span>
                    <span className={"article-bias " + bias.toLowerCase()}>
                        {bias}
                    </span>
                    {' | '}
                    <span className="article-date">
                        {formatDate(new Date(props.article.publishedTime))}
                    </span>
                </div>
            </div>
            <div className="right-container">
                <div className="sentiment-container">
                    Sentiment:
                    <div className={getSentimentClass()}>
                        {Math.round(props.article.textInfo.sentiment*100)}%
                    </div>
                </div>
                
                <div className="emotion-container">
                    Emotion:
                    <div className={getEmotionClass()}>
                        {capitalize(props.article.textInfo.emotion)}
                    </div>
                </div>
            </div>
        </a>
    )
}

File: backend/scheduled-function/article.py
from company import Company
from text_info import TextInfo


class Article:
    """Representation of online news article.

    Attributes:
        company (Company): Company name.
        title (str): Article title.
        description (str): First lines of article.
        article_url (str): Link to article.
        image_url (str): Link to article thumbnail.
        published_time (str): Time article was posted in ISO 8601 format
    """

    def __init__(self, company: Company, title: str, description: str,
                 published_time: str, article_url: str, image_url: str,
                 text_info: TextInfo = None):
        self.company = company
        self.title = title
        self.description = description
        self.published_time = published_time
        self.article_url = article_url
        self.image_url = image_url
        self.text_info = text_info

    def __str__(self):
        return str(self.company) + ', ' + \
               self.title[:10] + ', ' + \
               self.description[:10] + ', ' + \
               self.published_time + ', ' + \
               self.article_url + ', ' + \
               self.image_url

    def __hash__(self) -> int:
        return hash(self.article_url)

    def __eq__(self, o: object) -> bool:
        if isinstance(o, Article):
            return self.article_url == o.article_url
        return False
